pkg/importer/imageio-datasource.go

```diff
@@ -277,30 +497,36 @@ func getTransfer(conn ConnectionInterface, diskID string) (*ovirtsdk4.ImageTrans
 		return nil, uint64(0), errors.New("Error disk id not available")
 	}
 
-	image, err := ovirtsdk4.NewImageBuilder().Id(id).Build()
-	if err != nil {
-		return nil, uint64(0), errors.Wrap(err, "Error building image object")
+	var imageTransferBuilder *ovirtsdk4.ImageTransferBuilder
+	if snapshot != nil {
+		imageTransferBuilder = ovirtsdk4.NewImageTransferBuilder().Snapshot(snapshot)
+	} else {
+		image, err := ovirtsdk4.NewImageBuilder().Id(id).Build()
+		if err != nil {
+			return nil, uint64(0), errors.Wrap(err, "Error building image object")
+		}
+		imageTransferBuilder = ovirtsdk4.NewImageTransferBuilder().Image(image)
 	}
 
-	transfersService := conn.SystemService().ImageTransfersService()
-	transfer := transfersService.Add()
-	imageTransfer, err := ovirtsdk4.NewImageTransferBuilder().Image(
-		image,
-	).Direction(
+	imageTransfer, err := imageTransferBuilder.Direction(
 		ovirtsdk4.IMAGETRANSFERDIRECTION_DOWNLOAD,
 	).Format(
-		ovirtsdk4.DISKFORMAT_RAW,
+		formatType,
+	).InactivityTimeout(
+		60,
 	).Build()
 	if err != nil {
 		return nil, uint64(0), errors.Wrap(err, "Error preparing transfer object")
 	}
 
+	transfersService := conn.SystemService().ImageTransfersService()
+	transfer := transfersService.Add()
 	transfer.ImageTransfer(imageTransfer)
 	var it = &ovirtsdk4.ImageTransfer{}
 	for {
 		response, err := transfer.Send()
 		if err != nil {
-			if strings.Contains(err.Error(), "Disk is locked") {
+			if strings.Contains(err.Error(), "Disk is locked") || strings.Contains(err.Error(), "disks are locked") {
```

Comment:
Does transfer.Send() return a typed error? I am not a huge fan of checking for strings in an error if the source is sending a typed error.

Reply:
No, there is no typed error for disk locked unfortunately. This comes directly from the oVirt engine and there is no way to identify it properly. However, you can sidestep the issue by using the correlation_id trick. This involves setting the correlation_id query parameter and then checking if the jobs with that ID are completed. See here: https://github.com/oVirt/go-ovirt-client/blob/5f8e8c7a3b188698e905f54b40a091d64accc88b/client_util.go#L24

Reply:
This is a good trick, but the disk (or snapshot) is not always locked by our own job. If the disk is already locked, there's not much the importer can do except wait for it to unlock.

Reply:
Fair enough.

Reply:
Okay we should encourage the sdk authors to return some typed errors, so we don't have to search for strings in errors in order to determine stuff.